crates/weaver_semconv_gen/src/gen.rs

```diff
@@ -0,0 +1,476 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Markdown writing utilities.
+
+use crate::{Error, GenerateMarkdownArgs, ResolvedSemconvRegistry};
+use itertools::Itertools;
+use std::fmt::Write;
+use weaver_resolved_schema::attribute::Attribute;
+use weaver_resolved_schema::registry::Group;
+use weaver_semconv::attribute::{
+    AttributeType, BasicRequirementLevelSpec, EnumEntriesSpec, Examples, PrimitiveOrArrayTypeSpec,
+    RequirementLevel, TemplateTypeSpec, ValueSpec,
+};
+use weaver_semconv::group::{GroupType, InstrumentSpec};
+
+// The size a string is allowed to be before it is pushed into notes.
+const BREAK_COUNT: usize = 50;
+
+/// Context around the generation of markdown that we use to avoid conflicts
+/// between multiple templates within the same markdown file.
+#[derive(Default)]
+pub struct GenerateMarkdownContext {
+    /// The notes that have been added to the current markdown snippet.
+    notes: Vec<String>,
+}
+
+impl GenerateMarkdownContext {
+    /// Adds a note to the context and returns a link to its index.
+    fn add_note(&mut self, note: String) -> String {
+        self.notes.push(note);
+        let idx = self.notes.len();
+        format!("[{idx}]")
+    }
+
+    /// Renders stored notes into markdown format.
+    fn write_rendered_notes<Out: Write>(&self, out: &mut Out) -> Result<(), Error> {
+        for (counter, note) in self.notes.iter().enumerate() {
+            write!(out, "\n**[{}]:** {}\n", counter + 1, note.trim())?;
+        }
+        Ok(())
+    }
+}
+
+/// Determines an enum's type by the type of its values.
+fn enum_type_string(members: &Vec<EnumEntriesSpec>) -> &'static str {
```

Comment:
```suggestion
fn enum_type_string(members: &[EnumEntriesSpec]) -> &'static str {
```
Clippy should have complain about this one :-)

Reply:
Agreed.  Fixed